### PR TITLE
Add the HEATMAP line style.

### PR DESCRIPTION
**Pull request by manolama**
Add the HeatmapDef class used to configure heatmaps globaly for a graph. Add a palette option to the LineDef class for use by heatmaps. Parse the HeatmapDef in Grapher.scala and pass it to the Axis and then PlotDef.
Throw an exception if a user attempts to plot a heatmap and stacked lines on the same axis.
Select a color or palette for use in heatmaps in Grapher.scala. If no style or heatmap settings are provided, the heatmap grabs a color based on the order of the expression and uses alpha transparencies for the color scale. Remaining expressions use the rest of the palette as normal. If the user specifies a color for the expression, that color is used with alphas and no colors are taken from the main palette. If a palette is specified for heatmaps, that palette is used for the color scale.

### CODE REVIEW

File: atlas-chart/src/main/scala/com/netflix/atlas/chart/model/DataDef.scala [hunk 1]
```diff
@@ -60,7 +62,8 @@ case class LineDef(
   color: Color = Color.RED,
   lineStyle: LineStyle = LineStyle.LINE,
   lineWidth: Float = 1.0f,
-  legendStats: SummaryStats = SummaryStats.empty
+  legendStats: SummaryStats = SummaryStats.empty,
+  palette: Option[Palette] = None
```
Comment (brharrington) [inline atlas-chart/src/main/scala/com/netflix/atlas/chart/model/DataDef.scala:66]: Why is this needed? I would have expected the palette for a heatmap to be resolved before this and would be set on the HeatmapDef.

File: atlas-eval/src/main/scala/com/netflix/atlas/eval/graph/Grapher.scala [hunk 2]
```diff
@@ -295,6 +295,19 @@ case class Grapher(settings: DefaultSettings) {
...
+        if (
+          (dfltStyle == LineStyle.STACK || exprs.exists(
+            _.lineStyle.fold(dfltStyle)(s => LineStyle.valueOf(s.toUpperCase)) == LineStyle.STACK
+          )) && exprs.exists(
+            _.lineStyle
+              .fold(dfltStyle)(s => LineStyle.valueOf(s.toUpperCase)) == LineStyle.HEATMAP
+          )
+        ) {
+          throw new IllegalArgumentException(
+            "Mixing STACK and HEATMAP line styles on the same axis is not allowed."
```
Comment (brharrington) [inline atlas-eval/src/main/scala/com/netflix/atlas/eval/graph/Grapher.scala:307]: > Throw an exception if a user attempts to plot a heatmap and stacked lines on the same axis.

What is the rationale for this? While I would agree it is not generally going to be useful, I don't really see a reason to prohibit that combination.

**Review by brharrington — COMMENTED** [review]
Added a few questions after a first pass, will take a closer look in a bit.

File: atlas-chart/src/main/scala/com/netflix/atlas/chart/model/DataDef.scala (hunk 1 above)
Comment (manolama) [reply atlas-chart/src/main/scala/com/netflix/atlas/chart/model/DataDef.scala:66]: I could do it that way, via heatmap def. I just have to modify the `Axis` in the `Grapher`. Should be easy.

File: atlas-eval/src/main/scala/com/netflix/atlas/eval/graph/Grapher.scala (hunk 2 above)
Comment (manolama) [reply atlas-eval/src/main/scala/com/netflix/atlas/eval/graph/Grapher.scala:307]: For the percentile heatmap, I created a new ValueAxis type. If stacking is involved, the calculations pollute a lot of the heatmap logic (along with the axis) and I figured it wasn't worth it to support an op that could easily wind up completely obscuring the map. And if someone accidentally tried to plot percentiles or something else they wanted to line up with the map, it wouldn't so they'd be more confused. I think it's best to just ask the user to modify their query in that case.

**Comment by brharrington** [comment]
Incorporated as part of #1547.
